orderer/common/follower/follower_chain.go

```diff
@@ -426,6 +426,129 @@ func (c *Chain) pullUpToJoin() error {
 // It will exit with 'nil' if it detects a config block that indicates the orderer is a member of the cluster.
 // It checks whether the chain was stopped between blocks.
 func (c *Chain) pullAfterJoin() error {
-	//TODO
-	return errors.New("not implemented yet: pull after join block")
+	if c.lastConfig == nil {
+		err := c.loadLastConfig()
+		if err != nil {
+			return errors.Wrap(err, "failed to load last config block")
+		}
+	}
+
+	c.logger.Debugf("last config block: %d", c.lastConfig.Header.Number)
+	channelMember, err := c.clusterConsenter.IsChannelMember(c.lastConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to determine channel membership from last config")
+	}
+
+	for !channelMember {
+		configBlock, err := c.pullUntilNextConfig()
+		if err != nil {
+			return errors.Wrap(err, "failed to pull until next config")
+		}
+
+		channelMember, err = c.clusterConsenter.IsChannelMember(configBlock)
+		if err != nil {
+			return errors.Wrap(err, "failed to determine channel membership from pulled config")
+		}
+
+		c.logger.Debugf("next config block: %d, channel member: %v", c.lastConfig.Header.Number, channelMember)
+	}
+
+	return err
+}
+
+// pullUntilNextConfig will return the next config block or an error.
+func (c *Chain) pullUntilNextConfig() (*common.Block, error) {
+	// This puller is built from the tip of the ledger. When we get a new config block we exit the method. That will
+	// destroy the puller. Next time we enter, a new puller is built, taking in the new config.
+	puller, err := c.createBlockPullerFunc()
+	if err != nil {
+		c.logger.Errorf("Error creating block puller: %s", err)
+		return nil, errors.Wrap(err, "error creating block puller")
+	}
+	defer puller.Close()
+
+	var configBlock *common.Block
+	for configBlock == nil {
+		endpoint, latestHeight, err := cluster.LatestHeightAndEndpoint(puller)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get latest height and endpoint")
+		}
+		c.logger.Debugf("Orderer endpoint %s has the biggest ledger height: %d", endpoint, latestHeight)
+
+		if latestHeight <= c.ledgerResources.Height() {
+			c.logger.Debugf("My height: %d, latest height: %d; going to wait %v for latest height to grow",
+				c.ledgerResources.Height(), latestHeight, c.retryInterval)
+			select {
+			case <-c.stopChan:
+				return nil, errors.New("stopped while waiting for latest height to grow")
+			case <-c.timeAfter(c.retryInterval):
```

Comment:
This pattern really doesn't make sense to me here.

If we all the OSNs are up to date for this channel, say... there's no traffic on it, then we begin re-checking with the network for updates every 100ms by default, with an exponential backoff to 60s?  Why the exponential backoff? Why not simply pick a sensible refresh interval?

I'd also note, that the call to `cluster.LatestHeightAndEndpoint` is not necessarily a light one.  This hits all of the OSNs.

This is also a case where creating a common block puller path might make more sense.  Orderers are designed to deliver streams of blocks to peers as they are created, but here, we end up polling the network to look for updates.  We might want to poll periodically to make sure we're not connected to someone who is lagging, but generally, we should just be able to let the blocks arrive, like a peer.

Reply:
Let's start form the beginning... at the risk of iterating things you most probably already know... 
The cluster.BlockPuller available in the orderer was designed to do on-boarding or state transfer, that is, to efficiently pull blocks that we know exist. So for on-boarding until the join-block, it is good. However it was not designed to wait for the stream of blocks as the delivery client in the peer does. I am aware of that. There were 3 possibilities:
1. write a simple - maybe not so efficient - but working solution on top of cluster.BlockPuller
2. augment the cluster.BlockPuller to behave (also) like the delivery client in the peer
3. write a unified delivery client for both orderer and peer.

I chose (1) because I think that for the main use case we describe in the RFC, the period in which the follower pulls blocks past the join-block is limited in time. An admin will detect that on-boarding had finished (status changes from "onboarding" to "active"), and will re-configure the channel to include the orderer in the consenters set. This period is expected to be limited in time and relatively short. 

The scenario in which an orderer is following the cluster for a long period of time, like a "hot backup", is low priority at the moment, I think. If it turns out that it is a frequent use case, we may need to go to (2) or (3).

Now, with respect to the polling frequency. What is a reasonable frequency? It depends on the rate of block arrival, and on how costly it is to poll. We can certainly put a constant "reasonable number there, say Tpoll=500ms. The idea behind the backoff is to automatically explore and adapt to the approximate right number. Note that the retryInterval increases on consecutive failed attempts, but resets to the minimum on a successful attempt. On a channel with little traffic, say a block every T seconds, and a minimal polling interval of Tmin, we'll get a success every ~O(log(T/Tmin)), rather than ~(T/Tpoll). What we get is a kind of a saw-tooth  behavior that straddles the right interval. (Another option is to decrease on success, either dividing by the base or subtracting a constant).

Reply:
I added different min & max interval defaults for the case when the follower pulls past the join block. Grows from 500ms to 10s with base 1.5.

Reply:
The height polling policy and the pull retry policy were separated in the last commit. 
All retries are now in function scope for readability.